packages/components/src/internal/components/domainproperties/dataset/DatasetDesignerPanels.tsx

```diff
@@ -316,6 +323,34 @@ export class DatasetDesignerPanelImpl extends React.PureComponent<Props & Inject
         );
     };
 
+    onImportErrorStayAndFix = () => {
+        const { model } = this.state;
+
+        this.setState({importErrorModalOpen: false});
+        const dropOptions = {
+            schemaName: 'study',
+            queryName: model.name,
+            failure: (error) => {
+                this.setState(
+                    produce((draft: Draft<State>) => {
+                        draft.model.exception = error;
+                }))
+            },
+            success: () => {
+                this.setState(produce((draft: Draft<State>) => {
+                    draft.model.domain = this.state.preSaveDomain;
```

Comment:
How to set state based on existing state with Immer?

Reply:
This looks correct. Is it not working?

Reply:
Works fine, but I was wondering if there was an analogue to the `this.setState((state) => ({ <state assignments> }));` pattern where the body of the arrow function gets passed the existing state—although I don't think it's crucial here because we aren't mutating based on current state, I'd thought that that pattern generally can help prevent some finicky clicking-really-fast or other bugs